src/ctap/customization.rs

```diff
@@ -0,0 +1,247 @@
+
+/// Limits the number of RP IDs that can change the minimum PIN length.
+///
+/// # Invariant
+///
+/// - If this value is 0, DEFAULT_MIN_PIN_LENGTH_RP_IDS must be non-empty.
+///
+/// You can use this constant to have an upper limit in storage requirements.
+/// This might be useful if you want to more reliably predict the remaining
+/// storage. Stored string can still be of arbitrary length though, until RP ID
+/// truncation is implemented.
+/// Outside of memory considerations, you can set this value to 0 if only RP IDs
+/// in DEFAULT_MIN_PIN_LENGTH_RP_IDS should be allowed to change the minimum PIN
+/// length.
+pub const MAX_RP_IDS_LENGTH: usize = 8;
+
+/// Sets the number of resident keys you can store.
+///
+/// # Invariant
+///
+/// - The storage key CREDENTIALS must fit at least this number of credentials.
+///
+/// This value has implications on the flash lifetime, please see the
+/// documentation for NUM_PAGES below.
+pub const MAX_SUPPORTED_RESIDENT_KEYS: usize = 150;
+
+/// Sets the number of pages used for persistent storage.
+///
+/// The number of pages should be at least 3 and at most what the flash can
+/// hold. There should be no reason to put a small number here, except that the
+/// latency of flash operations is linear in the number of pages. This may
+/// improve in the future. Currently, using 20 pages gives between 20ms and
+/// 240ms per operation. The rule of thumb is between 1ms and 12ms per
+/// additional page.
+///
+/// Limiting the number of resident keys permits to ensure a minimum number of
+/// counter increments.
+/// Let:
+/// - P the number of pages (NUM_PAGES)
+/// - K the maximum number of resident keys (MAX_SUPPORTED_RESIDENT_KEYS)
+/// - S the maximum size of a resident key (about 500)
+/// - C the number of erase cycles (10000)
+/// - I the minimum number of counter increments
+///
+/// We have: I = (P * 4084 - 5107 - K * S) / 8 * C
+///
+/// With P=20 and K=150, we have I=2M which is enough for 500 increments per day
+/// for 10 years.
+pub const NUM_PAGES: usize = 20;
+
+#[allow(clippy::assertions_on_constants)]
+pub fn check_invariants() {
+    assert!(DEFAULT_MIN_PIN_LENGTH >= 4);
+    assert!(DEFAULT_MIN_PIN_LENGTH <= 63);
+    assert!(!USE_BATCH_ATTESTATION || ENTERPRISE_ATTESTATION_MODE.is_none());
+    if let Some(EnterpriseAttestationMode::VendorFacilitated) = ENTERPRISE_ATTESTATION_MODE {
+        assert!(!ENTERPRISE_RP_ID_LIST.is_empty());
+    }
+    assert!(MAX_PIN_RETRIES <= 8);
+    assert!(MAX_CRED_BLOB_LENGTH >= 32);
+    if let Some(count) = MAX_CREDENTIAL_COUNT_IN_LIST {
+        assert!(count >= 1);
+    }
+    assert!(MAX_LARGE_BLOB_ARRAY_SIZE >= 1024);
+    // Storage tests check if MAX_LARGE_BLOB_ARRAY_SIZE fits the shards.
+    if MAX_RP_IDS_LENGTH == 0 {
+        assert!(!DEFAULT_MIN_PIN_LENGTH_RP_IDS.is_empty());
+    }
+    // Storage keys tests check if MAX_SUPPORTED_RESIDENT_KEYS fits CREDENTIALS.
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
I'm pretty sure that nobody's going to run `cargo test` after changing the customization.
And if I remember our previous discussion, compile-time asserts are not supported yet without a crate.

This leaves us with one solution: can we run only this unit tests from `customization.rs` before we compile OpenSK app in `deploy.py`?

Reply:
Those seem to be check when the board boots (and crash it). But yeah, would be good to have something at compile time. We could run something like `cargo test --features=std --lib customization` or `cargo test --features=std --lib static_test` where all tests that should run statically contain `static_test` in their name.

Reply:
This PR does not change the current behaviour. It still panics when the configuration is wrong, just tests more. It additionally has a test to check the invariants. Adding this test to deploy sounds good. Question is if we still need to check invariants in `CtapState::new()` then?

Reply:
If broken invariants would have bad consequences (like corrupting data, leaking data, etc.) then we need them, if the device would otherwise crash without other bad consequences then we probably don't need them, but that's just optimization. Those tests look fast.

Reply:
Preventing compilation of a bad firmware seems to me a better behavior than a device crashing at boot (and potentially people starting opening issues). I'm fine with merging this PR until we agree :)

Reply:
IMO Usually all tests should pass before you deploy. The difference here is that we expect non-developers to change these values, so we want the extra protection. The test in main wasn't really the end-game solution anyway, so I went ahead and added
```rust
cargo test --features=std --lib customization
```

to `deploy.py`.